### PR TITLE
Added test for set evidence

### PR DESCRIPTION
**Pull request by Sekhar-Kumar-Dash**
## Changes proposed

Added tests for set_evidence.py

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

**Comment by AlyaGomaa** [comment]
you probably need to add the test_set_evidence.py file to  `.github/workflows/CI-production-testing.yml` and `.github/workflows/CI-staging.yml` to be able to run it in CI

File: .github/workflows/CI-production-testing.yml [hunk 1]
```diff
@@ -167,7 +167,11 @@ jobs:
         coverage run --source=./ -m pytest -s tests/integration_tests/test_portscans.py -p no:warnings -vv
         coverage report --include="modules/network_discovery/*"
         coverage html --include="modules/network_discovery/*" -d coverage_reports/network_discovery
-
+    
+    - name: set evidence test
+      run: |
+        coverage run --source=./ -m pytest tests/test_set_evidence.py -p no:warnings -vv
+
```
Comment (AlyaGomaa) [inline .github/workflows/CI-production-testing.yml:174]: this command is missing
coverage report --include=""
 coverage html --include="" -d 
 should be exactly like the 3 lines above

Comment (Sekhar-Kumar-Dash) [reply .github/workflows/CI-production-testing.yml:174]: okay

File: tests/test_set_evidence.py [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+    assert evidence.threat_level == ThreatLevel.MEDIUM
+    assert evidence.confidence == 0.5
+    assert evidence.category == IDEACategory.ANOMALY_TRAFFIC
+    assert evidence.profile.ip == "192.168.0.1"
+    assert evidence.timewindow.number == 2
+    assert sorted(evidence.uid) == sorted(["unique_id1", "unique_id2"])
+
+
+def test_connection_to_multiple_ports(set_evidence_helper, db_mock):
+    """Testing the connection_to_multiple_ports method."""
```
Comment (AlyaGomaa) [inline tests/test_set_evidence.py:540]: there's an if statement in this function that should be tested

File: tests/test_set_evidence.py [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+    assert evidence.threat_level == ThreatLevel.INFO
+    assert evidence.confidence == 0.7
+    assert evidence.category == IDEACategory.ANOMALY_BEHAVIOUR
+    assert evidence.profile.ip == "192.168.0.1"
+    assert evidence.timewindow.number == 5
+    assert evidence.uid == ["unique_id"]
+
+
+def test_for_port_0_connection(set_evidence_helper, db_mock):
+    """Testing the for_port_0_connection method."""
```
Comment (AlyaGomaa) [inline tests/test_set_evidence.py:626]: there's also and if statement here that should be tested

File: tests/test_set_evidence.py [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+    args, _ = db_mock.set_evidence.call_args_list[0]
+    evidence = args[0]
+    assert evidence.evidence_type == EvidenceType.MALICIOUS_JA3S
+    assert evidence.attacker.value == "10.0.0.1"
+    assert evidence.threat_level == ThreatLevel.HIGH
+    assert evidence.category == IDEACategory.INTRUSION_BOTNET
+    assert evidence.profile.ip == "10.0.0.1"
+    assert evidence.timewindow.number == 9
+    assert evidence.uid == ["unique_id"]
+    assert evidence.source_target_tag == Tag.CC
```
Comment (AlyaGomaa) [inline tests/test_set_evidence.py:682]: there were 2 evidence here and you only have assertions for one of them

File: tests/test_set_evidence.py [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+    assert evidence.threat_level == ThreatLevel.MEDIUM
+    assert evidence.category == IDEACategory.INTRUSION_BOTNET
+    assert evidence.profile.ip == "192.168.0.1"
+    assert evidence.timewindow.number == 10
+    assert evidence.uid == ["unique_id"]
+    assert evidence.source_target_tag == Tag.BOTNET
+
+
+def test_data_exfiltration(set_evidence_helper, db_mock):
+    """Testing the data_exfiltration method."""
```
Comment (AlyaGomaa) [inline tests/test_set_evidence.py:717]: there are 2 evidence here and you're only checking for one

File: tests/test_set_evidence.py [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+    assert evidence.threat_level == ThreatLevel.HIGH
+    assert evidence.category == IDEACategory.ATTEMPT_LOGIN
+    assert evidence.profile.ip == "192.168.0.1"
+    assert evidence.timewindow.number == 13
+    assert sorted(evidence.uid) == sorted(uid)
+    assert evidence.conn_count == smtp_bruteforce_threshold
+
+
+def test_malicious_ssl(set_evidence_helper, db_mock):
+    """Testing the malicious_ssl method."""
```
Comment (AlyaGomaa) [inline tests/test_set_evidence.py:794]: 2 evidence here and youre only checking for 1

**Review by AlyaGomaa — CHANGES_REQUESTED** [review]
Hello @sekhar thanks for the effort, we have a few missing test cases and i'd like to know why you're not using existing fixtures and module_factory objects and creating youir own instead?

File: tests/test_set_evidence.py [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+    """Testing the dns_without_conn method."""
+    result = set_evidence_helper.dns_without_conn(
+        domain="example.com",
+        timestamp="2023-05-06T12:00:00Z",
+        profileid="profile_192.168.0.1",
+        twid="timewindow1",
+        uid="unique_id",
+    )
+
+    if result is True:
```
[the remark this answers is not in the document]
Comment (Sekhar-Kumar-Dash) [reply tests/test_set_evidence.py:235]: I was focused on verifying that the evidence was set correctly, but I forgot to check the actual return value of the dns_without_conn method. It looks like I accidentally did this by looking at another function or copied some logic from another test and didn't adjust it properly. I think the correct value should be None.
No i haven't used chatgpt Actually, I have tried using Chatgpt, but it is terrible at generating tests for complex code, and it always fails the test although it does recommend some test cases.

File: tests/test_set_evidence.py (hunk 6 above)
Comment (Sekhar-Kumar-Dash) [reply tests/test_set_evidence.py:794]: okay I will make this changes and then update this PR

**Comment by Sekhar-Kumar-Dash** [comment]
@AlyaGomaa  sorry I did not push the changes i made to the actual code of set evidence  I will add this to my notes this wont happen again

**Comment by AlyaGomaa** [comment]
hey @Sekhar-Kumar-Dash can you take a look at this one?

**Comment by Sekhar-Kumar-Dash** [comment]
> hey @Sekhar-Kumar-Dash can you take a look at this one?

@AlyaGomaa  this looks strange are you sure these are related to my code as before PR only the unittest was failing and every test was failing and now it is just the opposite

**Comment by AlyaGomaa** [comment]
@Sekhar-Kumar-Dash let me check

**Pull request merged**